File: scripts/fix_index_html.py
#!/usr/bin/env python3

"""This program does various fixes on the `public/index.html` file."""

import sys # import system module
import re # import regular expression module
from fileinput import FileInput # import fileinput module

class FixIndexHtml:
  """
  This class contains all of the functionality for fixing the `public/index.html` file.
  """

  # create filepath class-property
  filepath = ''

  # define constructor
  def __init__(self, filepath):
    # initialize filepath
    self.filepath = filepath

  # define move_script_tags method
  def move_script_tags(self):
    """
    The purpose of this function is to:
      1. move the script tags from the `public/index.html` from the bottom to the top
      2. add the defer attributes to each one of those script tags
      3. add the `type="text/javascript"` mime-type attribute to script tags that don't have `type="module"`
      4. close of the link tag with the main stylesheet attached to it and add `type="text/css"`
    """

    print('Moving around script tags in index.html ... ')

    # regex for identifying script tags (constant)
    # use https://regexr.com for testing regular expressions
    SCRIPT_TAG_REGEX = re.compile(r'((?:<script src=\")(?:(?:runtime|polyfills|main|vendor|scripts)(?:-?(?:(?:es(?:(?:201)?[56789]))|(?:latest)))?)(?:\.)(?:(?:#?(?:[\da-fA-F]{2}){3}){3}(?:[\da-fA-F]){2})(?:\.js\")(?:(?:\stype="module")?(?:\snomodule)?)(?:><\/script>))')

    # future script tag string from joined script tag strings array
    new_script_tags = ''

    # wrap file I/O logic in try block
    try:

      # the first file I/O operation will store the script tags in a variable

      # open index.html file for reading
      index_file = open(self.filepath, 'r')

      # loop through each line in file
      for line in index_file:

        # store script tags in list
        script_tags = SCRIPT_TAG_REGEX.findall(line)

        # if script tags exist in line
        if script_tags:

          # add text/javascript mime type if type="module" does not exist in script tag
          for index, script_tag in enumerate(script_tags):
            if 'type="module"' not in script_tag:
              script_tags[index] = script_tag.replace('src="', 'type="text/javascript" src="')

          # set new script tags to joined array and add the defer attr to each script element
          new_script_tags = ('\n'.join(script_tags).replace('></script>', ' defer></script>')
                                                   .replace('type="module" ', '')
                                                   .replace('<script src="', '<script type="module" src="'))

      # close file from reading
      index_file.close()

      # the second file I/O operation wil move around the script tags regardless of formatting and close
      # off the stylesheet link tag and add its proper mime type

      # use fileinput to open up index.html file and create backup before editing
      with FileInput(self.filepath, inplace=True, backup='.bak') as file:

        # loop through each line in file
        for line in file:

          # store script tags in list variable
          script_tags = SCRIPT_TAG_REGEX.findall(line)

          # do various checks that depend on formatting

          # if the link tag and the closing head tag on the same line
          if ('.css">' in line or '<link rel="stylesheet" href="styles.' in line) and '</head>' in line:
            sys.stdout.write(line.replace('.css"></head>', f'.css" />\n\n{new_script_tags}\n</head>')
                                 .replace('rel="stylesheet"', 'rel="stylesheet" type="text/css"'))

          # if the link tag is not on the same line as the head tag
          elif '</head>' in line and ('.css">' not in line or '<link rel="stylesheet" href="styles.' not in line):
            sys.stdout.write(line.replace('</head>', f'{new_script_tags}\n</head>'))

          # if the head tag is not on the same line as the link tag (different condition)
          elif ('.css">' in line or '<link rel="stylesheet" href="styles.' in line) and '</head>' not in line:
            sys.stdout.write(line.replace('.css">', '.css" />\n')
                                 .replace('rel="stylesheet"', 'rel="stylesheet" type="text/css"'))

          # check if script tags are in current line
          elif script_tags:

            # loop thru script tags in bottom of file
            for i in range(len(script_tags)):
              # remove script tags from bottom of file
              line = line.replace(script_tags[i], '')

            # write changes
            sys.stdout.write(line)

          # otherwise...
          else:
            # print the other lines
            sys.stdout.write(line)

      # close the fileinput stream
      FileInput().close()

      print('Done!', end='\n\n')

    # catch file not found error
    except FileNotFoundError:
      print(f'Error, index.html file not found (looking in `{self.filepath}`.)', end='\n\n')

    # general exception
    except Exception as err:
      print(f'An error occurred:\n{str(err)}', end='\n\n')

  # define insert_comment method
  def insert_comment(self):
    """
    This function inserts a very important comment right below the `<!DOCTYPE html>` declaration in the
    `../public/index.html` file. It will also convert a lowercase `<!doctype html>` to an uppercase `<!DOCTYPE html>`
    if it needs to.
    """

    print('Inserting comment into index.html ... ')

    # use try block in case file is not found
    try:

      # define very important comment constant
      IMPORTANT_COMMENT = '<!-- May the source be with you! -->'

      # define vars
      text_to_search = '<!DOCTYPE html>'
      replacement_text = f'<!DOCTYPE html>\n{IMPORTANT_COMMENT}\n'

      # use fileinput to edit file
      with FileInput(self.filepath, inplace=True, backup='.bak1') as file:

        # loop thru each line in file
        for line in file:

          # add comment
          if text_to_search in line:

            # check if doctype declaration is on its own line
            if len(line.strip()) == 15:
              # if so, strip the line of any trailing and/or leading whitespace/tabs
              line = line.strip()
              # then insert the comment
              sys.stdout.write(line.replace(text_to_search, replacement_text))
            # otherwise
            else:
              # skip string trimming and insert comment
              sys.stdout.write(line.replace(text_to_search, replacement_text))

          # replace declaration and add comment
          elif text_to_search.lower() in line:
            line = line.strip()
            sys.stdout.write(line.replace(text_to_search.lower(), replacement_text))

          # print other lines
          else:
            sys.stdout.write(line)

      # close file stream
      FileInput().close()

      print('Done!', end='\n\n')

    # catch file not found error
    except FileNotFoundError:
      print(f'Error, index.html file not found (looking in `{self.filepath}`.)', end='\n\n')

    # catch general exception
    except Exception as err:
      print(f'An error occurred:\n{str(err)}', end='\n\n')

  # define add_positive_ssl_trusted_logo method
  def add_positive_ssl_trusted_logo(self):
    """
    This method adds the PositiveSSL trusted logo script to the bottom of the
    `../public/index.html` file as a comment.
    """

    print('Inserting PositiveSSL trusted logo as comment ... ')

    # define main const
    POSITIVE_SSL_HTML = (
                          '<!--\n'
                          '  <script type="text/javascript"> //<![CDATA[\n'
                          '  var tlJsHost = ((window.location.protocol == "https:") ? "https://secure.trust-provider.com/" : "http://www.trustlogo.com/");\n'
                          '  document.write(unescape("%3Cscript src=\'" + tlJsHost + "trustlogo/javascript/trustlogo.js\' type=\'text/javascript\'%\\3E%3C/script%\\3E"));\n'
                          '  //]]></script>\n'
                          '  <script language="JavaScript" type="text/javascript">\n'
                          '\tTrustLogo("https://www.positivessl.com/images/seals/positivessl_trust_seal_sm_124x32.png", "POSDV", "none");\n'
                          '  </script>\n'
                          '-->\n'
                        )

    # encapsulate file I/O logic in try-except block in case of errors
    try:

      # use fileinput to edit file
      with FileInput(self.filepath, inplace=True, backup='.bak2') as file:

        # loop thru each line in file
        for line in file:

          # check if line has body
          if '</head>' in line:
            # print the line
            print(line.replace('</head>', f'{POSITIVE_SSL_HTML}</head>'), end='')

          # otherwise...
          else:
            # print other lines
            print(line, end='')

      print('Done!', end='\n\n')

    # catch file not found error
    except FileNotFoundError:
      print(f'Error, index.html file not found (looking in `{self.filepath}`.)', end='\n\n')

    # catch general exception
    except Exception as err:
      print(f'An error occurred:\n{str(err)}', end='\n\n')

# apply fixes to index.html file
index_html = FixIndexHtml('../public/index.html')
index_html.move_script_tags()
index_html.insert_comment()
index_html.add_positive_ssl_trusted_logo()

File: scripts/FixIndexHTML.php
<?php
  namespace scripts;

  // reference Exception and Error namespaces
  use Exception;
  use Error;

  // check if file is being required by another file
  if (basename(__FILE__) != basename($_SERVER["SCRIPT_FILENAME"])) {
    // if so, throw an error
    throw new Error("Error: `FixIndexHTML.php` file cannot be required or included by another file.");
  }

  /**
   * Class FixIndexHTML
   * @package scripts
   */
  class FixIndexHTML
  {
    /**
     * @var null|string The file path of `index.html`.
     */
    protected $filepath = NULL;

    /**
     * @var null|string The new script tags (modified).
     */
    protected $new_script_tags = NULL;

    /**
     * @var null|string The current contents of the `index.html` file.
     */
    protected $contents = NULL;

    // create regular expression constants for detecting the script tags.
    const SCRIPT_TAG_REGEXP = "/((?:\x{003C}script src=\")(?:(?:runtime|polyfills(?:-es5)?|main|vendor|scripts)(?:(?:-)?(?:(?:es(?:(?:201)?[56789]))|(?:latest)))?)(?:\.)(?:[\da-fA-F]{20})(?:\.js\")(?:(?:\stype=\"module\")?(?:\snomodule)?(?:\sdefer)?)(?:\x{003E}\x{003C}\/script\x{003E}))/";
    const ASYNC_SCRIPT_TAG_REGEXP = "/((?:[\t\s]+)<script(?:[\t\s]+)type=[\"']text\/javascript[\"'](?:[\t\s]+)async(?:[\t\s]+)?>(?:\n+)?(?:[\t\s]+)?window\.Popper(?:[\t\s]+)?=(?:[\t\s]+)?\{(?:[\t\s]+)?\};?(?:\n+)?(?:[\t\s]+)?<\/script>)/";

    /**
     * FixIndexHTML constructor.
     * @param string $filepath The path to `index.html`.
     */
    function __construct($filepath)
    {
      // call method to set file path property.
      $this->set_filepath($filepath);
    }

    /**
     * @param string $filepath The file path to `index.html`.
     * @return void Nothing.
     */
    private function set_filepath($filepath)
    {
      // set class prop to method param.
      $this->filepath = $filepath;
    }

    public function backup_file()
    {
      // encapsulate main logic in try-catch
      try {
        // create file vars
        $backup_filepath = $this->filepath . ".bak";
        // get original index.html file contents
        $backup_file_contents = file_get_contents($this->filepath);
        // open file
        $file = fopen($backup_filepath, "w");
        // write to file
        fwrite($file, $backup_file_contents);
        // close file
        fclose($file);
      } catch (Exception $e) {
        // handle any exceptions
        echo "Error:\nCode: " . strval($e->getCode()) . "\nMessage: " . $e->getMessage();
        // stop script
        exit(1);
      }
    }

    /**
     * @return void Nothing.
     */
    public function read_script_tags()
    {
      // encapsulate main logic in try-catch (in case of errors)
      try {
        // create file variables
        // open file for reading
        $file = fopen($this->filepath, "r");
        $file_size = filesize($this->filepath);
        // set contents prop to text in file
        $this->contents = fread($file, $file_size);
        // output array var with all matching script tags (according to regexp)
        preg_match_all(self::SCRIPT_TAG_REGEXP, $this->contents, $script_tags);
        // override script tags var to first index of original regexp-match array
        $script_tags = $script_tags[0];
        // if script tags exist
        if ($script_tags != null && sizeof($script_tags) > 0) {
          // loop thru script tags (0th index)
          for ($i = 0; $i < count($script_tags); $i++) {
            // if type=module is absent from script tag
            if (strpos($script_tags[$i], "type=\"module\"") === false) {
              // add type=text/javascript
              $script_tags[$i] = str_replace("src=\"", "type=\"text/javascript\" src=\"", $script_tags[$i]);
            } else {
              // remove type="module" occurrence(s)
              $script_tags[$i] = preg_replace("/((?:\s?)type=\"module\"(?:\s?))+/", "", $script_tags[$i]);
              // add type=module right before src attr
              $script_tags[$i] = str_replace("src=\"", "type=\"module\" src=\"", $script_tags[$i]);
            }
            // if defer attribute is not present on script tag
            if (strpos($script_tags[$i], "defer") === false) {
              // add the defer attribute
              $script_tags[$i] = str_replace("></script>", " defer></script>", $script_tags[$i]);
            }
          }

          // move type=module to beginning of script tag and join script tags array into string (glued by newline character)
          $glued_script_tags = implode("\n", $script_tags);
          // if no type=module or type=text/javascript is detected, then add it
          # $glued_script_tags = str_replace("<script src=\"", "<script type=\"module\" src=\"", $glued_script_tags);
          // set new script tags prop to glued script tag's var
          $this->new_script_tags = $glued_script_tags; # str_replace("type=\"module\" ", "", $glued_script_tags);
        }
        // close the file
        fclose($file);
      } catch (Exception $e) {
        // handle any exceptions
        echo "Error:\nCode: " . strval($e->getCode()) . "\nMessage: " . $e->getMessage();
        // stop script
        exit(1);
      }
    }

    /**
     * @return void Nothing.
     */
    public function move_script_tags()
    {
      // encapsulate main logic in try-catch (in case of errors)
      try {
        // open file for writing
        $file = fopen($this->filepath, "w");
        // cache async script tag
        preg_match(self::ASYNC_SCRIPT_TAG_REGEXP, $this->contents, $async_script_tag);
        // set var as first index in array
        $async_script_tag = $async_script_tag[0];
        // get rid of async script tag
        $this->contents = str_replace($async_script_tag, "", $this->contents);
        // create file lines var by splitting contents by newline char
        $file_lines = explode("\n", $this->contents);
        // output array var with all matching script tags (according to regexp)
        preg_match_all(self::SCRIPT_TAG_REGEXP, $this->contents, $script_tags);
        // create boolean var for checking if script tags exist or not
        $script_tags_exist = ($script_tags != null && sizeof($script_tags) > 0) ? true : false;
        // loop thru each line in file
        foreach ($file_lines as $line) {
          // add in the script tags depending on where the main stylesheet is linked,
          // and do various checks that depend on the formatting of the file
          // if the link tag and the closing head tag on the same line
          if ((strpos($line, ".css\">") !== false || strpos($line, "<link rel=\"stylesheet\" href=\"styles.") !== false) && strpos($line, "</head>") !== false) {
            $updated_link_el_attrs = str_replace("rel=\"stylesheet\"", "rel=\"stylesheet\" type=\"text/css\"", $line);
            $updated_head_tags = str_replace(".css\"></head>", ".css\" />\n \n<!--[if !IE]><!-->\n" . $async_script_tag . "\n" . $this->new_script_tags . "\n<!--<![endif]-->\n</head>", $updated_link_el_attrs);
            fwrite($file, $updated_head_tags . "\n");
            // if the link tag is not on the same line as the head tag
          } elseif (strpos($line, "</head>") !== false && (strpos($line, ".css\">") === false || strpos($line, "<link rel=\"stylesheet\" href=\"styles.") === false)) {
            fwrite($file, str_replace("</head>", $async_script_tag . "\n" . $this->new_script_tags . "\n</head>", $line) . "\n");
            // if the head tag is not on the same line as the link tag (different condition)
          } elseif ((strpos($line, ".css\">") !== false || strpos($line, "<link rel=\"stylesheet\" href=\"styles.") !== false) and strpos($line, "</head>") === false) {
            $updated_link_el_attrs = str_replace("rel=\"stylesheet\"", "rel=\"stylesheet\" type=\"text/css\"", $line);
            $updated_link_tag_closed = str_replace(".css\">", ".css\" />\n", $updated_link_el_attrs);
            fwrite($file, $updated_link_tag_closed . "\n");
            // check if script tags are in current line
          } elseif ($script_tags_exist === true) {
            foreach ($script_tags[0] as $script_tag) {
              $line = str_replace($script_tag, "", $line);
            }
            fwrite($file, $line . "\n");
          } else {
            // print out every other line (followed by a newline)
            fwrite($file, $line . "\n");
          }
        }
        // close the file
        fclose($file);
      } catch (Exception $e) {
        // handle any exceptions
       echo "Error:\nCode: " . strval($e->getCode()) . "\nMessage: " . $e->getMessage();
       // kill the script
       exit(1);
      }
    }

    /**
     * @return void Nothing.
     */
    public function insert_comment()
    {
      // set content before main logic
      $this->set_content();
      // encapsulate main logic in try-catch (in case of errors)
      try {
        // define constant if not already defined
        if (!defined("IMPORTANT_COMMENT")) {
          define("IMPORTANT_COMMENT", "<!-- May the source be with you! -->");
        }
        // create file vars
        // open file for writing
        $file = fopen($this->filepath, "w");
        // create file lines var by splitting contents by newline char
        $file_lines = explode("\n", $this->contents);
        // define search and replace string vars
        $text_to_search = "<!DOCTYPE html>";
        $replacement_text = $text_to_search . "\n" . IMPORTANT_COMMENT . "\n";
        // loop thru each line in file
        foreach ($file_lines as $line) {
          // add comment logic
          if (strpos($line, $text_to_search) !== false) {
            // check if doctype declaration is on its own line
            if (strlen(self::strip($line)) == strlen($text_to_search)) {
              // if so, strip the line of any trailing and/or leading whitespace/tabs
              $line = self::strip($line);
              // then insert the comment
              fwrite($file, str_replace($text_to_search, $replacement_text, $line));
            } else {
              // otherwise just insert the comment
              fwrite($file, str_replace($text_to_search, $replacement_text, $line));
            }
            // replace lowercase doctype declaration with uppercase and add comment
          } elseif (strpos($line, strtolower($text_to_search)) !== false) {
            $line = self::strip($line);
            fwrite($file, str_replace(strtolower($text_to_search), $replacement_text, $line));
          } else {
            // else write every other line
            fwrite($file, $line . "\n");
          }
        }
        // close the file
        fclose($file);
      } catch (Exception $e) {
        // handle any exceptions
        echo "Error:\nCode: " . strval($e->getCode()) . "\nMessage: " . $e->getMessage();
        // kill the script
        exit(1);
      }
    }

    /**
     * @return void Nothing.
     */
    private function set_content()
    {
      // encapsulate main logic in try-catch (in case of errors)
      try {
        // set contents prop to text in file
        $this->contents = file_get_contents($this->filepath);
        // replace trailing newlines with blank string
        $this->contents = preg_replace("/\n+$/m", "", $this->contents);
      } catch (Exception $e) {
        // handle any exceptions
        echo "Error:\nCode: " . strval($e->getCode()) . "\nMessage: " . $e->getMessage();
        // stop the script
        exit(1);
      }
    }

    /**
     * @param string $string The string to strip whitespace and tabs from.
     * @return string|string[]|null The stripped string.
     */
    private static function strip($string) {
      // define regexp constants if not already defined
      if (!defined("LEFT_TRIM")) {
        define("LEFT_TRIM", "/^[\s\t]+/");
      }
      if (!defined("RIGHT_TRIM")) {
        define("RIGHT_TRIM", "/[\s\t]+$/");
      }
      // replace left and right leading/trailing whitespace and tabs on string with nothing.
      $strip_right = preg_replace(RIGHT_TRIM, "", $string);
      $strip_left = preg_replace(LEFT_TRIM, "", $strip_right);
      // convert to string just in case of hacking
      $stripped_string = strval($strip_left);
      // return stripped value
      return $stripped_string;
    }
  }

  /**
   * @return void Nothing.
   */
  function main() {
    // instantiate class and pass in path to `index.html` file
    $fix_index_html = new FixIndexHTML("../public/index.html");
    // first backup the file
    echo "Backup up index.html file ... \n";
    $fix_index_html->backup_file();
    echo "Done! (file saved as index.html.bak)\n\n";
    // begin fixing index.html file ...
    echo "Moving around script tags in index.html ... \n";
    $fix_index_html->read_script_tags();
    $fix_index_html->move_script_tags();
    echo "Done!\n\n";
    echo "Inserting comment into index.html ... \n";
    $fix_index_html->insert_comment();
    echo "Done!\n\n";
    // done!
  }

  // invoke main function
  main();
